server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1992,70 +1989,66 @@ protected final Releasable acquireEnsureOpenRef() {
-                return Thread.currentThread().getName().contains(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME)
-                    || super.blockingAllowed();
-            }
-        };
-        drainOnCloseListener.addListener(future);
-        try {
-            future.get();
-            return true;
-        } catch (ExecutionException e) {
-            logger.error("failure while draining operations on close", e);
-            assert false : e;
-            throw new IllegalStateException(e);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            logger.error("interrupted while draining operations on close");
-            throw new IllegalStateException(e);
-        }
+        drainOnCloseListener.addListener(listener);
     }
 
     /**
      * Flush the engine (committing segments to disk and truncating the translog) and close it.
      */
     public void flushAndClose() throws IOException {
+        // TODO Do we have to wait here?
+        flushAndClose(ActionListener.noop());
+    }
+
+    private void flushAndClose(ActionListener<Void> listener) throws IOException {
         logger.trace("flushAndClose() maybe draining ops");
-        if (isClosed.get() == false && drainForClose()) {
-            logger.trace("flushAndClose drained ops");
-            try {
-                logger.debug("flushing shard on close - this might take some time to sync files to disk");
+        if (isClosed.get() == false) {
+            drainForClose(listener.map(unused -> {
+                logger.trace("flushAndClose drained ops");
                 try {
-                    // TODO: We are not waiting for full durability here atm because we are on the cluster state update thread
-                    flushHoldingLock(false, false, ActionListener.noop());
-                } catch (AlreadyClosedException ex) {
-                    logger.debug("engine already closed - skipping flushAndClose");
+                    logger.debug("flushing shard on close - this might take some time to sync files to disk");
+                    try {
+                        // TODO: We are not waiting for full durability here atm because we are on the cluster state update thread
+                        flushHoldingLock(false, false, ActionListener.noop());
+                    } catch (AlreadyClosedException ex) {
+                        logger.debug("engine already closed - skipping flushAndClose");
+                    }
+                    return null;
+                } finally {
+                    closeNoLock("flushAndClose", closedLatch);
                 }
-            } finally {
-                closeNoLock("flushAndClose", closedLatch);
-            }
+            }));
         }
         awaitPendingClose();
     }
 
     @Override
     public void close() throws IOException {
+        close(ActionListener.noop());
```

Comment:
I don't think we can make this a fire-and-forget action. At best, it's very trappy that this method would return before the engine is closed.

Do we really need `Engine implements Closeable` anyway? If we dropped that we could make `close()` properly async.

Reply:
> Do we really need Engine implements Closeable anyway? If we dropped that we could make close() properly async.

Thank you! I was stuck on this point, too. Let me see whether it would be possible to get rid of the sync `flushAndClose` and `close` methods.